Feature: Implemented DAPO

lavaman131 opened the pull request:
We just finished implementing DAPO and applied it to training small language models, specifically with `unsloth/Qwen2.5-0.5B-Instruct` [here](https://github.com/lavaman131/reasoning-slms) (includes DAPO config usage) and were able to get some positive results as described in my [technical blog](https://alexlavaee.me/projects/reasoning-slms/).

I thought it might be useful to submit a PR in case the TRL community would benefit from a DAPO implementation. I tried to follow the rough structure of the GRPO trainer, but am of course happy to get any feedback on the code.

Here is a rough overview of the proposed API extending off of what the current GRPO API was missing:

* Soft Overlong Punishment

```python
use_soft_overlong_punishment: Optional[bool] = field(
    default=True,
    metadata={
        "help": "A length-aware penalty mechanism designed to shape the reward for truncated samples. According to the "
        "[DAPO](https://huggingface.co/papers/2503.14476) paper, this signals to the model to avoid excessively long responses."
    },
)
soft_punish_cache: Optional[int] = field(
    default=0,
    metadata={
        "help": "The L_cache term in the soft overlong punishnment calculation. According to the "
        "[DAPO](https://huggingface.co/papers/2503.14476) paper, this term determines how close"
        "to the max completion length a response can get before being penalized."
    },
)
```

* Dynamic Sampling

```python
use_dynamic_sampling: Optional[bool] = field(
    default=True,
    metadata={
        "help": "Whether to use dynamic sampling to filter out prompts that have zero variance from the "
        "[DAPO](https://huggingface.co/papers/2503.14476) paper."
    },
)
```

LeonEricsson commented:
The current implementation already supports overlong punishment through a custom reward function, so I’m not sure a separate DAPO trainer is necessary just for dynamic sampling. It's more practical+maintainable to integrate that functionality into the existing GRPO trainer. Also, have to performed any wall clock benchmarks for dynamic sampling? It certainly reduces the number of optimization steps but I'm not convinced it significantly improves training time.

lavaman131 commented:
Thanks for the feedback @LeonEricsson.

Regarding your questions:

1. I wasn't able to find an implementation of soft overlong punishment in the documentation and also not in [open_r1](https://github.com/huggingface/open-r1/blob/main/src/open_r1/rewards.py) where @shirinyamani and @qgallouedec are making some changes. I implemented the token-level soft-overlong punishment as described in the original paper. To my knowledge, the reward function API was not yet modified to take the `completion_ids` and was proposed [here](https://github.com/huggingface/trl/pull/3272) by @qgallouedec to support this. So, I took the proposed changes there and added it to my current implementation.

I agree, this doesn't need to be its own API, but once the new reward function callable is supported I'm happy to add a short tutorial in the docs page that uses the GRPOConfig with how to extend it with the custom reward function callable that takes `completion_ids` and dynamic sampling to make it DAPO.

2. Regarding performance benchmarks I've attached a screenshot of my wandb dashboard that compares vanilla GRPO to adding dynamic sampling during training. As of now, I currently implement dynamic sampling by making the loss zero, effectively skipping an optimization step. Is this the recommended way to implement this?
<img width="2185" alt="Screenshot 2025-05-05 at 9 18 06 AM" src="https://github.com/user-attachments/assets/8941a79e-eb6c-464e-bdf5-906ab4d924c6" />

shirinyamani commented:
Hi @lavaman131 thanks for your contribution. 
I think we have overlong_filtering support in this PR #3248 where we defined an arg to filter the overlong completions.

LeonEricsson commented:
> 1. I wasn't able to find an implementation of soft overlong punishment in the documentation and also not in [open_r1](https://github.com/huggingface/open-r1/blob/main/src/open_r1/rewards.py) where @shirinyamani and @qgallouedec are making some changes. I implemented the token-level soft-overlong punishment as described in the original paper. To my knowledge, the reward function API was not yet modified to take the `completion_ids` and was proposed [here](https://github.com/huggingface/trl/pull/3272) by @qgallouedec to support this. So, I took the proposed changes there and added it to my current implementation.

Aah right I see, I think its best to wait for #3272 and then implement overlong punishment in [open_r1](https://github.com/huggingface/open-r1/blob/main/src/open_r1/rewards.py). There's actually already an open [PR](https://github.com/huggingface/open-r1/pull/567) for this, the last thing discussed was a comparisons of computing soft_overlong_punishment on characters vs tokens.

> 2. Regarding performance benchmarks I've attached a screenshot of my wandb dashboard that compares vanilla GRPO to adding dynamic sampling during training. As of now, I currently implement dynamic sampling by making the loss zero, effectively skipping an optimization step. Is this the recommended way to implement this?

Hmm, this doesn’t sound quite right, though I might be misunderstanding. Dynamic sampling aims to improve sample efficiency by replacing zero-advantage samples with more informative ones. As training progresses, the number of samples where all generations are correct (or incorrect) tends to increase. These samples don’t contribute any gradients during policy updates, which leads to inefficiency.

To address this, DAPO proposes **over-sampling** - taking advantage of the fact that generation is highly parallelizable - to ensure that the batch is filled with non-zero advantage samples. This could involve continuously drawing new samples until the batch meets this criterion. Alternatively, one might naively increase the batch size by some factor and then filter out the zero-advantage samples, but that would require tuning said factor, which isn’t trivial.

lavaman131 commented:
Ok, thank you both for confirming.

**Notes:**
Yes, I was aware of the overlong filtering, but was referring to the soft overlong punishment from #3272 @shirinyamani.

Then, is it correct to assume that the soft overlong punishment should be implemented as a custom reward function since `completion_ids` is now accepted to the reward callable?

Regarding dynamic sampling @LeonEricsson, my way of preventing the zero-advantage samples was to mask them out and if there are no samples skip the optimization step. Right, I was kind-of doing the latter, where I was using more optimization steps, rather than increasing the batch size. I like your idea of continuously sampling until the batch meets this criterion, however, actually thought it would be messier calling `vllm.generate`.

**Takeaways:**
Based off of this discussion, would the community benefit from a page in the documentation, describing the custom reward function for soft overlong filtering and maybe even adding a flag for dynamic sampling if it is practical to do so in the first place (e.g. due to possible sample inefficiency).

qgallouedec commented:
> Then, is it correct to assume that the soft overlong punishment should be implemented as a custom reward function since `completion_ids` is now accepted to the reward callable?

I think so yes.

> would the community benefit from a page in the documentation, describing the custom reward function for soft overlong filtering

To take a broader view, I think we could even start to take the reward functions from open-r1 and integrate them into trl as a sub-module. And consequently this sub-module would have its own dedicated documentation page.

> and maybe even adding a flag for dynamic sampling if it is practical to do so in the first place (e.g. due to possible sample inefficiency).

I don't really know what to do about dynamic sampling. For the moment, I don't see any evidence that it would make a difference in compute efficiency, and it would introduce significant complexity into the code. I don't know if there's much more to be done in that direction.

lavaman131 commented:
Thanks, @qgallouedec, that makes sense.

I especially like the idea of a reward functions submodule from open-r1. I could create a draft PR proposing this functionality and would be happy to contribute with these efforts if agreeable?

qgallouedec commented:
OFC!!

qgallouedec commented:
> I especially like the idea of a reward functions submodule from open-r1

initial structure in #3430
